aspnetcore/security/cors.md

```diff
@@ -42,110 +42,97 @@ These URLs have different origins than the previous two URLs:
 
 ## Enable CORS with attributes
 
-The [&lbrack;EnableCors&rbrack;](xref:Microsoft.AspNetCore.Cors.EnableCorsAttribute) attribute provides an alternative to applying CORS globally. The `[EnableCors]` attribute enables CORS for selected end points, rather than all end points.
+Enabling CORS with the [[EnableCors]](xref:Microsoft.AspNetCore.Cors.EnableCorsAttribute) attribute to only those endpoints that require CORS is preferred to enabling CORS globally.
 
-Use `[EnableCors]` to specify the default policy and `[EnableCors("{Policy String}")]` to specify a policy.
+The [[EnableCors]](xref:Microsoft.AspNetCore.Cors.EnableCorsAttribute) attribute provides an alternative to applying CORS globally. The `[EnableCors]` attribute enables CORS for selected endpoints, rather than all endpoints:
+
+* `[EnableCors]` specifies the default policy.
+* `[EnableCors("{Policy String}")]` specifies a specific a policy.
 
 The `[EnableCors]` attribute can be applied to:
 
 * Razor Page `PageModel`
 * Controller
 * Controller action method
 
-You can apply different policies to controller/page-model/action with the  `[EnableCors]` attribute. When the `[EnableCors]` attribute is applied to a controllers/page-model/action method, and CORS is enabled in middleware, both policies are applied. We recommend against combining policies. Use the `[EnableCors]` attribute or middleware, not both in the same app.
+Different policies can be applied to controllers, page models, or action methods with the `[EnableCors]` attribute. When the `[EnableCors]` attribute is applied to a controller, page model, or action method, and CORS is enabled in middleware, ***both*** policies are applied. ***We recommend against combining policies. Use the*** `[EnableCors]` ***attribute or middleware, not both in the same app.***
 
 The following code applies a different policy to each method:
 
-[!code-csharp[](cors/sample/Cors/WebAPI/Controllers/WidgetController.cs?name=snippet&highlight=6,14)]
+[!code-csharp[](cors/3.1sample/Cors/WebAPI/Controllers/WidgetController.cs?name=snippet&highlight=6,14)]
 
 The following code creates a CORS default policy and a policy named `"AnotherPolicy"`:
 
-[!code-csharp[](cors/sample/Cors/WebAPI/StartupMultiPolicy.cs?name=snippet&highlight=12-28)]
+[!code-csharp[](cors/3.1sample/Cors/WebAPI/Startup3.cs?name=snippet&highlight=12-28,44)]
+
+See [Test CORS](#testc) for instructions on testing code similar to the preceding code.
 
 ### Disable CORS
 
-The [&lbrack;DisableCors&rbrack;](xref:Microsoft.AspNetCore.Cors.DisableCorsAttribute) attribute disables CORS for the controller/page-model/action.
+The [[DisableCors]](xref:Microsoft.AspNetCore.Cors.DisableCorsAttribute) attribute does ***not***  disables CORS when a CORS policy has been enabled in middleware. For example, either of the following calls in `Startup.Configure` sets a CORS policy in middleware:
```

Comment:
`[DisableCors]` _does_ override the middleware setup when using endpoint routing, but not when using whatever-the-other-routing-was-called.

Reply:
I added a H2 near the top that lists the 3 approaches with links to each section:

There are three ways to enable CORS:

    In middleware using a named policy or default policy.
    Using endpoint routing.
    With the [EnableCors] attribute.

Testing the deployed app, From https://cors1.azurewebsites.net/ which calls 
Request URL: https://cors3.azurewebsites.net/api/values/GetValues2
`[DisableCors]` fails. The cors3 server uses:

` endpoints.MapControllers().RequireCors(MyPolicy);`

Testing from https://cors3.azurewebsites.net/ which calls
Request URL: https://cors1.azurewebsites.net/api/values/GetValues2
correctly blocks CORS with the following error:
Access to fetch at 'https://cors1.azurewebsites.net/api/values/GetValues2' from origin 'https://cors3.azurewebsites.net' has been blocked by CORS policy:

Cors1 deployment uses StartupTest

```
 public void ConfigureServices(IServiceCollection services)
        {
            services.AddCors(options =>
            {
                options.AddDefaultPolicy(
                    builder =>

...

app.UseCors();

            app.UseAuthorization();
            app.UseAuthorization();

            app.UseEndpoints(endpoints =>
            {
                endpoints.MapControllers();
                endpoints.MapRazorPages();
            });
```

Reply:
@serpent5  can you review the public link and this change.

Reply:
@serpent5  hold off, I'm still working on your next comment. It seems if you're going to use `[EnableCors]`, it doesn't make sense with the default policy. If you defined a default policy, you still must call  `app.UseCors()` - which enables the default policy for all endpoints.

So I think I need to change the sample to define "MyPolicy", then call  `app.UseCors()` 
That will take me some time to do that and test everything.

Reply:
Right, so we were talking about slightly different things. To me, middleware meant something applied with either a default or named policy using `AddCors`/`UseCors`. That's option 1 on your list above. With that, `[DisableCors]` _does_ work. With option 2, using `RequireCors`, `[DisableCors]` does _not_ work, as you said.

It gets a bit weird if you do something like this:

```csharp
[EnableCors("Policy")]
[DisableCors]
public IActionResult SomeAction()
```

In this case, `[DisableCors]` wins, assuming just those two attributes. However, consider this:

```csharp
[DisableCors]
[EnableCors("Policy")]
public IActionResult SomeAction()
```

In this case, `EnableCors("Policy")` wins. This is just because the endpoint metadata is ordered, and of course it's a weird example, but I found it interesting anyway.

So, it looks like `[EnableCors(...)]` and `[DisableCors]` have no effect if `RequireCors` is used. I guess that's intentional, but I find it weird.

Reply:
> It seems if you're going to use `[EnableCors]`, it doesn't make sense with the default policy

I could see a case where you might like to set up a default policy and override it using `[EnableCors]`. You can't do that if you use `RequireCors` for setting up the default policy, but you could do something like this:

```csharp
services.AddCors(options =>
{
    options.AddDefaultPolicy(builder => builder.WithOrigins("http://origin/one"));
    options.AddPolicy("SomethingElse", builder => builder.WithOrigins("http://origin/two"));
});

---

[EnableCors("SomethingElse")]
public IActionResult SomeAction()
```

With just that setup, all endpoints use the default, except for the `SomeAction` endpoint.